### PR TITLE
New command `dart pub cache preload`

### PR DESCRIPTION
**Pull request by sigurdm**
Will load one or more tar.gz archives of packages as those served by pub.get and add them to the cache.

Will overwrite an existing entry for the same package-version.

No attempt is made to verify this against what is really being served -> this can be done offline.
Writes the sha256 hash of the archive as the content-hash in the cache -> later pub get invocations will verify the cached hash against what is in the cache.

Fixes: https://github.com/dart-lang/pub/issues/3532

### CODE REVIEW

File: lib/src/command/cache_preload.dart [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  String get name => 'preload';
+  @override
+  String get description => 'Install packages from a .tar.gz archive.';
+  @override
+  String get argumentsDescription => '<package1.tar.gz> ...';
+  @override
+  String get docUrl => 'https://dart.dev/tools/pub/cmd/pub-cache';
+
+  @override
+  bool get hidden => true;
```
Comment (jonasfj) [inline lib/src/command/cache_preload.dart:25]:
```suggestion
  /// The `cache preload` command is hidden by default, because it's really only intended for
  /// `flutter` to use when pre-loading `PUB_CACHE` after being installed from `zip` archive.
  @override
  bool get hidden => true;
```

File: lib/src/command/cache_preload.dart [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import 'dart:async';
+
+import '../command.dart';
+import '../io.dart';
+import '../log.dart' as log;
+import '../source/hosted.dart';
+import '../utils.dart';
+
+/// Handles the `cache add` pub command.
```
Comment (jonasfj) [inline lib/src/command/cache_preload.dart:13]:
```suggestion
/// Handles the `cache preload` pub command.
```

File: lib/src/source/hosted.dart [hunk 3]
```diff
@@ -1086,9 +1086,14 @@ See $contentHashesDocumentationUrl.
...
-      await extractTarGz(readBinaryFileAsStream(archivePath), tempDir);
+      try {
+        await extractTarGz(readBinaryFileAsStream(archivePath), tempDir);
 
-      ensureDir(p.dirname(destPath));
+        ensureDir(p.dirname(destPath));
+      } catch (e) {
+        deleteEntry(tempDir);
+        rethrow;
+      }
```
Comment (jonasfj) [inline lib/src/source/hosted.dart:1096]: Nice, unrelated fix, right? :rofl:

File: test/cache/preload_test.dart [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+    await appDir({'foo': '^2.0.0'}).create();
+    // Do a `pub get` here to create a lock file in order to validate we later can
+    // `pub get --offline` with packages installed by `preload`.
+    await pubGet();
+
+    await runPub(args: ['cache', 'clean', '-f']);
+
+    final archivePath = p.join(sandbox, 'archive');
+    final archivePath2 = p.join(sandbox, 'archive2');
```
Comment (jonasfj) [inline test/cache/preload_test.dart:30]:
```suggestion
    final archivePath1 = p.join(sandbox, 'foo-1.0.0-archive.tar.gz');
    final archivePath2 = p.join(sandbox, 'foo-2.0.0-archive.tar.gz');
```

File: test/cache/preload_test.dart [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+    File(archivePath).writeAsBytesSync(await readBytes(
+        Uri.parse(server.url).resolve('packages/foo/versions/1.0.0.tar.gz')));
+    File(archivePath2).writeAsBytesSync(await readBytes(
+        Uri.parse(server.url).resolve('packages/foo/versions/2.0.0.tar.gz')));
```
Comment (jonasfj) [inline test/cache/preload_test.dart:35]: nit: but are you sure you don't want a trailing comma here :rofl: (do what you like, I can't really care :see_no_evil: )
```suggestion
    File(archivePath).writeAsBytesSync(
      await readBytes(Uri.parse(server.url).resolve('packages/foo/versions/1.0.0.tar.gz')),
    );
    File(archivePath2).writeAsBytesSync(
      await readBytes(Uri.parse(server.url).resolve('packages/foo/versions/2.0.0.tar.gz')),
    );
```

File: test/cache/preload_test.dart [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    final archivePath = p.join(sandbox, 'archive');
+
+    File(archivePath).writeAsBytesSync(await readBytes(
+        Uri.parse(server.url).resolve('packages/foo/versions/1.0.0.tar.gz')));
+    await runPub(
+      args: ['cache', 'preload', archivePath],
+      environment: {'_PUB_TEST_DEFAULT_HOSTED_URL': 'pub.dev'},
+      output: allOf([
+        contains(
+            'Installed $archivePath in cache as foo 1.0.0 from ${server.url}.')
```
Comment (jonasfj) [inline test/cache/preload_test.dart:73]: I don't understand this test case?

**Review by jonasfj — APPROVED** [review]

File: lib/src/source/hosted.dart (hunk 3 above)
Comment (sigurdm) [reply lib/src/source/hosted.dart:1096]: Yes :)

File: test/cache/preload_test.dart [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    final server = await servePackages();
+    server.serve('foo', '1.0.0');
+
+    final archivePath = p.join(sandbox, 'archive');
+
+    File(archivePath).writeAsBytesSync(await readBytes(
+        Uri.parse(server.url).resolve('packages/foo/versions/1.0.0.tar.gz')));
+    await runPub(
+      args: ['cache', 'preload', archivePath],
+      environment: {'_PUB_TEST_DEFAULT_HOSTED_URL': 'pub.dev'},
```
Comment (jonasfj) [inline test/cache/preload_test.dart:70]:
```suggestion
      // '${server.url}' is the default value for _PUB_TEST_DEFAULT_HOSTED_URL in tests
      // by overwriting it, we make the pub command think '${server.url}' is a custom pub server.
      environment: {'_PUB_TEST_DEFAULT_HOSTED_URL': 'pub.dev'},
```

**Review by jonasfj — APPROVED** [review]

File: test/cache/preload_test.dart (hunk 7 above)
Comment (sigurdm) [reply test/cache/preload_test.dart:75]: Added a comment along those lines

**Comment by parlough** [comment]
Sounds like we shouldn't document this on dart.dev, is that correct?

**Comment by sigurdm** [comment]
> Sounds like we shouldn't document this on dart.dev, is that correct?

Yeah for now we keep this unannounced mainly for Flutter use.
